rs/bitcoin/kyt/src/state.rs

```diff
@@ -0,0 +1,179 @@
+use crate::GetTxError;
+use bitcoin::{Address, Transaction};
+use ic_btc_interface::Txid;
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+
+#[derive(Debug, Clone)]
+pub enum FetchTxStatus<T> {
+    PendingOutcall,
+    PendingRetry { buffer_size: u32 },
+    Error(GetTxError),
+    Fetched(FetchedTx<T>),
+}
+
+#[derive(Debug, Clone)]
+pub struct FetchedTx<T> {
+    pub tx: T,
+    /// The index is the index of the input of tx, and the value is
+    /// the calculated address when the input has been fetched.
+    pub input_addresses: Vec<Option<Address>>,
+}
+
+// Max number of concurrent http outcalls.
+const MAX_CONCURRENT: u32 = 100;
+
+// The internal KYT state includes:
+// 1. Outcall capacity, a semaphore limiting max concurrent outcalls.
+// 2. fetch transaction status, indexed by transaction id.
+//
+// TODO(XC-191): persist canister state
+thread_local! {
+    static OUTCALL_CAPACITY: RefCell<u32> = const { RefCell::new(MAX_CONCURRENT) };
+    static FETCH_TX_STATUS: RefCell<BTreeMap<Txid, FetchTxStatus<Transaction>>> = RefCell::new(BTreeMap::default());
+}
+
+pub fn get_fetch_status(txid: Txid) -> Option<FetchTxStatus<Transaction>> {
+    FETCH_TX_STATUS.with(|s| s.borrow().get(&txid).cloned())
+}
+
+pub fn set_fetch_status(txid: Txid, status: FetchTxStatus<Transaction>) {
+    FETCH_TX_STATUS.with(|s| {
+        let _ = s.borrow_mut().insert(txid, status);
+    })
+}
+
+pub fn clear_fetch_status(txid: Txid) {
+    FETCH_TX_STATUS.with(|s| {
+        let _ = s.borrow_mut().remove(&txid);
+    })
+}
+
+/// Set the address at the given `index` in the `Fetched` status of the given `txid`.
+/// Pre-condition: the status of `txid` is `Fetched`, and `index` is within bounds.
+pub fn set_fetched_address(txid: Txid, index: usize, address: Address) {
+    FETCH_TX_STATUS.with(|s| {
+        s.borrow_mut().entry(txid).and_modify(|status| {
+            if let FetchTxStatus::Fetched(fetched) = status {
```

Comment:
While you do check the (pre-)condition that the status is `Fetched`, there is no check that `index` is not too large. Why is that?

Reply:
Usually pre-condition is something that should not be explicitly checked inside the function. The caller needs to make sure this is the case, and if needed, the pre-condition should be checked at the call site. 

Another way to think of this is, if such conditions are checked inside the function and handled accordingly, they are not actually pre-conditions, because the function is already accepting inputs that violate the pre-conditions.

So the status of `Fetched` is also not checked here. It is merely being matched, and we don't handle or throw error when it not matched.